packages/hoppscotch-cli/src/utils/getters.ts

```diff
@@ -134,3 +152,98 @@ export const roundDuration = (
+ * @param {string} params.pathOrId - The path to the local file or the ID for remote retrieval.
+ * @param {string} [params.accessToken] - The access token for authorizing remote retrieval.
+ * @param {string} [params.serverUrl] - The SH instance server URL for remote retrieval. Defaults to the cloud instance.
+ * @param {"collection" | "environment"} params.resourceType - The type of the resource to retrieve.
+ * @returns {Promise<HoppCollection | Environment>} A promise that resolves to the contents of the resource.
+ * @throws Will throw an error if the content type of the fetched resource is not `application/json`,
+ *         if there is an issue with the access token, if the server connection is refused,
+ *         or if the server URL is invalid.
+ */
+export const getResourceContents = async (
+  params: GetResourceContentsParams
+): Promise<HoppCollection | Environment> => {
+  const { pathOrId, accessToken, serverUrl, resourceType } = params;
+
+  let contents: HoppCollection | Environment | null = null;
+  let fileExistsInPath = false;
+
+  try {
+    await fs.access(pathOrId);
+    fileExistsInPath = true;
+  } catch (e) {
+    fileExistsInPath = false;
+  }
+
+  if (accessToken && !fileExistsInPath) {
+    try {
+      const hostname = serverUrl || "https://api.hoppscotch.io";
+
+      const separator = hostname.endsWith("/") ? "" : "/";
+      const resourcePath =
+        resourceType === "collection" ? "collection" : "environment";
+
+      const url = `${hostname}${separator}v1/access-tokens/${resourcePath}/${pathOrId}`;
+
+      const { data, headers } = await axios.get(url, {
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+        },
+      });
+
+      if (!headers["content-type"].includes("application/json")) {
+        throw new Error("INVALID_CONTENT_TYPE");
+      }
+
+      contents =
+        resourceType === "collection"
+          ? transformWorkspaceCollection(data as WorkspaceCollection)
+          : transformWorkspaceEnvironment(data as WorkspaceEnvironment);
+    } catch (err) {
+      const axiosErr = err as AxiosError<{
+        reason?: "TOKEN_EXPIRED" | "TOKEN_INVALID" | "INVALID_ID";
+        message: string;
+        statusCode: number;
+      }>;
+
+      const errReason = axiosErr.response?.data?.reason;
+
+      if (errReason) {
+        throw error({
+          code: errReason,
+          data: ["TOKEN_EXPIRED", "TOKEN_INVALID"].includes(errReason)
+            ? accessToken
+            : pathOrId,
+        });
+      }
+
+      if (axiosErr.code === "ECONNREFUSED") {
+        throw error({ code: "SERVER_CONNECTION_REFUSED", data: serverUrl });
+      }
+
+      if (
+        axiosErr.message === "INVALID_CONTENT_TYPE" ||
+        axiosErr.code === "ERR_INVALID_URL" ||
+        axiosErr.code === "ENOTFOUND" ||
+        axiosErr.response?.status === 404
+      ) {
+        throw error({ code: "INVALID_SERVER_URL", data: serverUrl });
```

Comment:
if the user does not provide a server URL and the api.hoppscotch.io server returned some error, `serverURL` will be `undefined`

for eg, right now since the changes are not in api.hoppscotch.io, it will 404 and we'll get this error message, `Please provide a valid SH instance server URL: undefined`

If we move the `serverURL` to the top of the try block, it won't be undefined, it will be the user provided value or the default api.hoppscotch.io.

the error message will be more meaningful this way.